### ISSUE TITLE
Configuration option HAVE_SET_TRIGGER not exposed

### ISSUE THREAD

**Issue by Ivan-Velickovic** [seL4/seL4#1073]
When making systems that abstract over seL4 system calls/invocations (e.g. seL4 Core Platform) or have more build-time checks (e.g the seL4 Rust crate), it would be useful to be able to check whether a platform's IRQ controller has the ability to set the trigger of an IRQ before making an invocation such as `seL4_IRQControl_GetTrigger`.

**Comment by Indanz** [comment]
`HAVE_SET_TRIGGER` is not a configuration option, even though the code checks it with `config_set()`. It is defined by the IRQ controller code used for the platform. You could check indirectly for value of `INTERRUPT_CONTROLLER`, though that's a bit awkward.

`seL4_IRQControl_GetTriggerCore` should only be available on SMP builds though, is there any platform that supports SMP, but not setting the IRQs core? Potential candidates are am335x, bcm2837 and omap3. There is also a plic-dummy for riscv which does define `HAVE_SET_TRIGGER`, used by ariane, rocketchip and spike platforms.

(Also, does anyone know an easy way to check whether a platform supports SMP?)

I agree that giving compile time errors instead of runtime errors is preferable.

**Comment by Ivan-Velickovic** [comment]
> `HAVE_SET_TRIGGER` is not a configuration option, even though the code checks it with `config_set()`. It is defined by the IRQ controller code used for the platform.

Yes I just called it a configuration option in the general sense that it is an option for IRQ controllers to configure... But that's confusing so I've updated the issue.

> (Also, does anyone know an easy way to check whether a platform supports SMP?)

`CONFIG_ENABLE_SMP_SUPPORT`?

**Comment by Ivan-Velickovic** [comment]
Right now I don't think there is actually any restriction on building an SMP for a platform, even if there is only one CPU defined in the DTS. I just tried building the kernel with `KernelMaxNumNodes=4` with only one CPU node in the DTS and it still built.

**Comment by lsf37** [comment]
> Right now I don't think there is actually any restriction on building an SMP kernel for a platform, even if there is only one CPU defined in the DTS. I just tried building the kernel with `KernelMaxNumNodes=4` with only one CPU node in the DTS for the `imx8mm-evk` and the kernel still built.

Hm, that's something we should probably fix. Does it just crash when you boot it or is something in the build system resetting the number of CPUs to 1?

**Comment by kent-mcleod** [comment]
Something like `KernelArmIRQCanSetTrigger`?

**Comment by Ivan-Velickovic** [comment]
> Something like `KernelArmIRQCanSetTrigger`?

Yes, just not sure what the best way to go about it in the seL4 build system. One approach would be to have each platform's `config.cmake` set the option but that does not seem ideal.

**Comment by Ivan-Velickovic** [comment]
> > Something like `KernelArmIRQCanSetTrigger`?
> 
> Yes, just not sure what the best way to go about it in the seL4 build system. One approach would be to have each platform's `config.cmake` set the option but that does not seem ideal.

I guess we could have it set to true by default and the platforms that do not support it can explicitly turn it off in `config.cmake` (I think there is only a few platforms that don't support setting the trigger).

**Comment by kent-mcleod** [comment]
I think it's essentially whether the platform uses the GIC (either version).  There should already be a config option for that in the build somewhere, so in arch/arm/config.cmake you could set this new option using config_set if the other option condition matches.

**Comment by Ivan-Velickovic** [comment]
> I think it's essentially whether the platform uses the GIC (either version). There should already be a config option for that in the build somewhere, so in arch/arm/config.cmake you could set this new option using config_set if the other option condition matches.

Currently I only see a way to check if GICv3 is used with `KernelArmGicV3`.
